tsl/src/bgw_policy/continuous_aggregate_api.c

```diff
@@ -39,3 +54,236 @@ policy_continuous_aggregate_proc(PG_FUNCTION_ARGS)
+		{
+			elog(WARNING,
+				 "could not add refresh policy due to existing policy on continuous aggregate with "
+				 "different arguments");
+			PG_RETURN_INT32(-1);
+		}
+	}
+
+	/* Next, insert a new job into jobs table */
+	namestrcpy(&application_name, "Refresh Policy");
+	namestrcpy(&refresh_name, "refresh");
+	namestrcpy(&proc_name, POLICY_REFRESH_CAGG_PROC_NAME);
+	namestrcpy(&proc_schema, INTERNAL_SCHEMA_NAME);
+	namestrcpy(&owner, GetUserNameFromId(owner_id, false));
+
+	if (IS_INTEGER_TYPE(dim_type) &&
+		!(IS_INTEGER_TYPE(start_interval_type) && IS_INTEGER_TYPE(end_interval_type)))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid parameter value for start_interval or end_interval"),
+				 errhint("integer time duration is required for continuous aggregates with integer "
+						 "time bucket ")));
+	}
+	if (IS_TIMESTAMP_TYPE(dim_type) &&
+		!((start_interval_type == INTERVALOID) && (end_interval_type == INTERVALOID)))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid parameter value for start_interval or end_interval"),
+				 errhint("interval time duration is required for continuous aggregate with "
+						 "timestamp-based time bucket ")));
+	}
+
+	JsonbParseState *parse_state = NULL;
+	pushJsonbValue(&parse_state, WJB_BEGIN_OBJECT, NULL);
+	ts_jsonb_add_int32(parse_state, CONFIG_KEY_MAT_HYPERTABLE_ID, mat_htid);
+	json_add_dim_interval_value(parse_state,
+								CONFIG_KEY_START_INTERVAL,
+								start_interval_type,
+								start_interval);
+	json_add_dim_interval_value(parse_state,
+								CONFIG_KEY_END_INTERVAL,
+								end_interval_type,
+								end_interval);
+	JsonbValue *result = pushJsonbValue(&parse_state, WJB_END_OBJECT, NULL);
+	Jsonb *config = JsonbValueToJsonb(result);
+
+	job_id = ts_bgw_job_insert_relation(&application_name,
+										&refresh_name,
+										&refresh_interval,
+										DEFAULT_MAX_RUNTIME,
+										DEFAULT_MAX_RETRIES,
+										&refresh_interval,
+										&proc_name,
+										&proc_schema,
+										&owner,
+										true,
+										mat_htid,
+										config);
+
+	PG_RETURN_INT32(job_id);
+}
+
+Datum
+policy_refresh_cagg_remove(PG_FUNCTION_ARGS)
+{
+	Oid cagg_oid = PG_GETARG_OID(0);
+	bool if_exists = PG_GETARG_BOOL(1);
+	int32 mat_htid;
+
+	ContinuousAgg *cagg = ts_continuous_agg_find_by_relid(cagg_oid);
+	if (!cagg)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("\"%s\" is not a continuous aggregate", get_rel_name(cagg_oid))));
+	}
+
+	ts_cagg_permissions_check(cagg_oid, GetUserId());
+	mat_htid = cagg->data.mat_hypertable_id;
+	List *jobs = ts_bgw_job_find_by_proc_and_hypertable_id(POLICY_REFRESH_CAGG_PROC_NAME,
+														   INTERNAL_SCHEMA_NAME,
+														   mat_htid);
+	if (jobs == NIL)
+	{
+		if (!if_exists)
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_OBJECT),
+					 errmsg("cannot remove refresh policy, no such policy exists")));
```

Comment:
I suggest making the error messages similar for the skipping and non-skipping case:

```suggestion
					 errmsg("refresh policy does not exist on continuous aggregate \"%s\"", get_rel_name(cagg_oid))));
```

Reply:
modified